### PR TITLE
Pattern inspector

### PR DESCRIPTION
**Pull request by milg15**
TODO: 

- [x] Change Image 
- [x] Rename relevant files
- [x] Set "initializer" image

### CODE REVIEW

**Comment by milg15** [comment]
There are some errors in the preview but it should be fix on the next issue when we change the previews to use the Canvas instead of css.

File: packages/app/src/components/inspector/FillInputFieldWithPicker.tsx
Comment (dabbott) [inline packages/app/src/components/inspector/FillInputFieldWithPicker.tsx:87]: Can we call `useApplicationState` here instead of passing these as props?

File: packages/app/src/components/inspector/GradientInspector.tsx
Comment (dabbott) [inline packages/app/src/components/inspector/GradientInspector.tsx:89]: I don't think we want to delete the `onNudgeOpacity` case here, since that'll be important when we support modifying fills for multiple layers, all with different colors. In that case, we wouldn't want to override the entire color. Probably same with `onChangeOpacity` above

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (dabbott) [inline packages/app/src/components/inspector/PatternInspector.tsx:198]: We have a `useHover` hook which we can use for this that handles some pointer weirdness. See:

https://github.com/noya-app/noya/blob/60d6525289203f0a325f805e7621ab50dc899261/packages/noya-designsystem/src/components/ListView.tsx#L178

We can pass a `onHoverChange` function to this hook to get updates about hover state

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (dabbott) [inline packages/app/src/components/inspector/PatternInspector.tsx:140]: We'll probably want a mapping from type => extension, both to show an error for invalid file types, and to handle jpeg. E.g.

```ts
const SUPPORTED_FILE_TYPES = {
  'image/png': 'png',
  'image/jpeg': 'jpg',
  'image/webp': 'webp'
}

// ...

const extension = SUPPORTED_FILE_TYPES[file.type]

if (!extension) {
  alert(`Files of type ${file.type} aren't supported. Files of type png, jpg, and webp are supported.`)
}
```

File: packages/noya-designsystem/src/utils/sketchPattern.ts
Comment (dabbott) [inline packages/noya-designsystem/src/utils/sketchPattern.ts:6]: I don't think we need to handle `ico` or `x-icon` files. Did you encounter this case?

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (dabbott) [reply packages/app/src/components/inspector/PatternInspector.tsx:140]: If we still need `mimeTypes` or `extensions` as an array, we can do `Object.keys(SUPPORTED_FILE_TYPES)` or `Object.values(SUPPORTED_FILE_TYPES)`

File: packages/noya-designsystem/src/utils/sketchPattern.ts
Comment (dabbott) [inline packages/noya-designsystem/src/utils/sketchPattern.ts:16]: Since we have this in a separate file already, should we delete it here? Presumably we'll end up replacing this with CanvasKit API calls anyway, so not super important

File: packages/noya-designsystem/src/utils/sketchPattern.ts
Comment (dabbott) [inline packages/noya-designsystem/src/utils/sketchPattern.ts:62]: Maybe a `switch` would be clearer?

File: packages/noya-state/src/reducers/styleReducer.ts
Comment (dabbott) [inline packages/noya-state/src/reducers/styleReducer.ts:322]: Maybe we should only do this if the `type` is pattern, just so we don't create objects we don't need in the save data. Same with gradient above

File: packages/noya-state/src/reducers/styleReducer.ts
Comment (dabbott) [inline packages/noya-state/src/reducers/styleReducer.ts:400]: We can use `clamp` here

File: packages/noya-state/src/reducers/styleReducer.ts
Comment (dabbott) [inline packages/noya-state/src/reducers/styleReducer.ts:423]: Also `clamp` here

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (dabbott) [inline packages/app/src/components/inspector/PatternInspector.tsx:163]: Should this be `if (on !== undefined)`? Otherwise `setIsDragOver` will not be called for `false`

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (dabbott) [inline packages/app/src/components/inspector/PatternInspector.tsx:168]: Let's try separating the drag and drop handling into a `hooks/useFileDropTarget.ts`.

```ts
function useFileDropTarget() {
  // useState stuff
  return {
    dropTargetProps: {
      onDragOver: ...,
      onDragEnter: ...,
      onDragLeave: ...,
      onDrop: ...
    },
    isDropTargetActive, // maybe instead of `isDropOver`
  }
}
```

Then this component can use it:

```tsx
const { dropTargetProps, isDropTargetActive } = useFileDropTarget()

return (
  // ...
    <Container {...dropTargetProps} isActive={isDropTargetActive}>...</Container>
  // ...
)
```

File: packages/app/src/components/inspector/FillRow.tsx
Comment (dabbott) [inline packages/app/src/components/inspector/FillRow.tsx:137]: I think I prefer it like this, since otherwise we'll have to ignore certain types in the callbacks of border/shadow pickers

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (dabbott) [inline packages/app/src/components/inspector/PatternInspector.tsx:76]: Do these need to be optional? Probably if these callbacks don't exist we shouldn't render this component within the parent at all

**Review by dabbott — CHANGES_REQUESTED** [review]
🔥 This looks great! I found a few things we can try refactoring before we move on to the canvas portion of this.

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (milg15) [reply packages/app/src/components/inspector/PatternInspector.tsx:76]: this is the same issue I have with gradients. All the actions are options because the are optional in the FillPicker. It's there a way of telling it that if the value._class it's gradient and patterns, those functions exist for sure ? 🤔

File: packages/noya-designsystem/src/utils/sketchPattern.ts
Comment (milg15) [reply packages/noya-designsystem/src/utils/sketchPattern.ts:6]: If we just leave it to the extensions we have. I'll remove that part  👍

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (dabbott) [reply packages/app/src/components/inspector/PatternInspector.tsx:76]: I think so... l'll see if I can come up with something tomorrow morning

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (milg15) [reply packages/app/src/components/inspector/PatternInspector.tsx:163]: ohhhhh!! That explains a lot! right 😆

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (milg15) [reply packages/app/src/components/inspector/PatternInspector.tsx:76]: I would do something similar to the onChanges. I think that way we won't have that many props in some components

Comment (dabbott) [reply packages/app/src/components/inspector/PatternInspector.tsx:76]: I tried improving this, but couldn't come up with anything great. The approach you suggested of using `value._class` look like it isn't supported yet but will be soon https://github.com/microsoft/TypeScript/pull/38839

I did see a few other things I want to try refactoring though, after we merge this

Comment (milg15) [reply packages/app/src/components/inspector/PatternInspector.tsx:76]: My brain it's in the future on this one 😆 I'll work on the other changes while I think of a solution for this

File: packages/noya-designsystem/src/utils/sketchPattern.ts
Comment (milg15) [reply packages/noya-designsystem/src/utils/sketchPattern.ts:16]: delete the encodeImg funcion? I just have it for preview for now. I'll delete it when we change to use CanvasKit

Comment (dabbott) [reply packages/noya-designsystem/src/utils/sketchPattern.ts:16]: Is this the same as what's in `utils/encodeImg.ts`? Or is it different?

Comment (milg15) [reply packages/noya-designsystem/src/utils/sketchPattern.ts:16]: Oh i forgot about that file 😅 I don't think we need it

File: packages/app/src/components/inspector/GradientInspector.tsx
Comment (milg15) [reply packages/app/src/components/inspector/GradientInspector.tsx:89]: 🤔 okey!

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (milg15) [reply packages/app/src/components/inspector/PatternInspector.tsx:77]: I work on everything but this part

File: packages/app/src/components/inspector/GradientInspector.tsx
Comment (dabbott) [reply packages/app/src/components/inspector/GradientInspector.tsx:93]: We should probably still put the nudge cases back in here and above

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (dabbott) [inline packages/app/src/components/inspector/PatternInspector.tsx:33]: This is clever, but will break if we change the color to a hex code or something. Let's add a new theme color for this overlay color

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (dabbott) [inline packages/app/src/components/inspector/PatternInspector.tsx:175]: This looks like it will call `handleImageFile` for each file in a list, but we probably only want to handle one. We could do:

```ts
const file = e.dataTransfer.items[0]

if (!file) return 
      
const entry = ...;
```

We should probably also show some user feedback with `alert` if the `kind` is wrong or we can't get the file as an `entry`

**Review by dabbott — APPROVED** [review]
Now it's working for me! I dunno what it was... anyway, looks great!

File: packages/app/src/hooks/useFileDropTarget.ts
Comment (dabbott) [inline packages/app/src/hooks/useFileDropTarget.ts:3]: Can `dropEvent` return just `void` here?

File: packages/app/src/components/inspector/PatternInspector.tsx
Comment (milg15) [reply packages/app/src/components/inspector/PatternInspector.tsx:175]: What should it say when we have the wrong kind or entry?

Comment (dabbott) [reply packages/app/src/components/inspector/PatternInspector.tsx:175]: - For wrong kind, "Only files are supported"
- For invalid entry, "Unable to read file ${filename}"

Comment (milg15) [reply packages/app/src/components/inspector/PatternInspector.tsx:175]: We don't use kind anymore, I change the to files instead of items. Because items it's used mostly when you want to also allow to drop directories and I don't think we are interested in that.

Comment (milg15) [reply packages/app/src/components/inspector/PatternInspector.tsx:175]: I can't get the file name if the file it's invalid 😅

**Comment by milg15** [comment]
I think with those changes it should be good to go 👍

**Pull request merged**
